### PR TITLE
 Ignore lock node during orphan connector node cleanup

### PR DESCRIPTION
**Pull request by vmaheshw**
Lock node will not be assigned to any instance and the orphan node cleanup code should ignore deleting this node by considering it as orphan node.

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java [hunk 1]
```diff
@@ -60,11 +60,15 @@ private KeyBuilder() {
    */
   private static final String DATASTREAM_TASK_CONFIG = CONNECTOR + "/%s/config";
 
+  /**
+   * lock node
+   */
+  protected static final String DATASTREAM_TASK_LOCK_ROOT_NAME = "lock";
```
Comment (ahmedahamid) [inline datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java:66]: Did you mean to make it package private? That's less accessible than `protected` and would still allow you to access it in `ZkAdapter`.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 2]
```diff
@@ -758,32 +765,37 @@ public void testDeleteTasksWithPrefix() {
...
 
     leftOverTasks = zkClient.getChildren(KeyBuilder.connector(testCluster, connectorType));
-    Assert.assertEquals(leftOverTasks.size(), 2);
+    Assert.assertEquals(leftOverTasks.size(), 3);
 
     adapter.cleanUpOrphanConnectorTasks(true);
 
     leftOverTasks = zkClient.getChildren(KeyBuilder.connector(testCluster, connectorType));
-    Assert.assertEquals(leftOverTasks.size(), 0);
+    Assert.assertEquals(leftOverTasks.size(), 1);
```
Comment (ahmedahamid) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:793]: Should we also assert that one znode left is actually `lock`?

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 3]
```diff
@@ -758,32 +765,37 @@ public void testDeleteTasksWithPrefix() {
...
+
+    // lock root node does not get deleted, once created.
+    adapter.releaseTask(lockTask);
+    leftOverTasks = zkClient.getChildren(KeyBuilder.connector(testCluster, connectorType));
+    Assert.assertEquals(leftOverTasks.size(), 1);
```
Comment (ahmedahamid) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:798]: This part seems a little extraneous. 

Should this test really assert that `lock` remains after `releaseTask()`? All it cares about is that `lock` survives `cleanUpOrphanConnectorTasks()`. On the other hand, if we change `ZkAdapter` in the future to have it delete `lock` if it's no longer needed, why should this test start failing then?

Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:798]: I feel this part is okay to have. If the behavior changes in the future, whoever is making the change will get alerted and get aware of this existing behavior ( which is not very obvious) and help in either fixing the test or their new behavior.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java (hunk 2 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:793]: done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java [hunk 4]
```diff
@@ -60,11 +60,15 @@ private KeyBuilder() {
    */
   private static final String DATASTREAM_TASK_CONFIG = CONNECTOR + "/%s/config";
 
+  /**
+   * lock node
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java]: done

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 5]
```diff
@@ -758,32 +765,37 @@ public void testDeleteTasksWithPrefix() {
...
+    Assert.assertEquals(leftOverTasks.size(), 3);
 
     adapter.cleanUpOrphanConnectorTasks(true);
 
     leftOverTasks = zkClient.getChildren(KeyBuilder.connector(testCluster, connectorType));
-    Assert.assertEquals(leftOverTasks.size(), 0);
+    Assert.assertEquals(leftOverTasks.size(), 1);
+
+    // lock root node does not get deleted, once created.
+    adapter.releaseTask(lockTask);
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:797]: the releaseTask deletes the connector/lock/<task>. The node connector/lock will not get deleted.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java (hunk 1 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/KeyBuilder.java]: done

**Pull request merged**